File: bot/bot.py
#!/usr/bin/env python
"""Bot module"""

import json
import asks
import trio
import trio_websocket

asks.init(trio)

def get_cookie_from_response(response, cookie_name):
    """Returns cookie value of cookie_name"""
    return next((c.value for c in response.cookies if c.name == cookie_name), None)

AUTH_COOKIE = 'QUART_AUTH'
DEBUG = False

class User:
    """User class"""
    def __init__(self, name: str="") -> None:
        self.name = name
        self.world = 0
        self.x = 0.0
        self.y = 0.0
        self.z = 0.0
        self.roll = 0.0
        self.yaw = 0.0
        self.pitch = 0.0
        self.avatar = 0
        self.state = 'idle'
        self.gesture = None

    def set_position(self, x: float = 0.0, y: float = 0.0, z: float = 0.0,
                     roll: float = 0.0, yaw: float = 0.0, pitch: float = 0.0,):
        self.x = x
        self.y = y
        self.z = z
        self.roll = roll
        self.yaw = yaw
        self.pitch = pitch

class Bot(User):
    """Bot class"""
    def __init__(self, web_url: str, ws_url: str, logging_enabled: bool=True) -> None:
        super().__init__()
        self.web_url = web_url
        self.ws_url = ws_url
        self.ws = None
        self.nursery = None
        self.logging_enabled = logging_enabled
        self.connected = False
        self.handlers = {}
        self.userlist = {}
        self.worldlist = {}
        self.cookiejar = {}

    def log(self, txt: str) -> None:
        if self.logging_enabled:
            print(txt)

    async def _callback(self, name: str, *parameters) -> None:
        for inst in [self] + list(self.handlers.values()):
            f = getattr(inst, name, None)
            if not hasattr(f, '__call__'):
                continue
            if DEBUG:
                self.log('calling %s() on instance %r' % (name, inst))
            if f is not None:
                await f(*parameters)

async def _process_msg(self, msg: dict) -> None:
    self.log(f"> {msg}")

    # Return early if the message doesn't have a 'type' field
    if 'type' not in msg:
        self.log("* unknown message")
        return

    t = msg['type']

    # Handle the different message types
    if t == "avatar":
        user = self.userlist.get(msg["user"])
        if user is not None:
            user.avatar = msg["data"]
        await self._callback("on_user_avatar", msg["user"], msg["data"])
    elif t == "join":
        await self._callback("on_user_join", msg["data"])
    elif t == "list":
        self.userlist.clear()
        for u in msg["data"]:
            user = User(u["name"])
            user.avatar = u["avatar"]
            user.world = u["world"]
            self.userlist[u["id"]] = user
        await self._callback("on_user_list")
    elif t == "msg":
        await self._callback("on_msg", msg["user"], msg["data"])
    elif t == "part":
        await self._callback("on_user_part", msg["data"])
    elif t == "pos":
        user = self.userlist.get(msg["user"])
        if user is not None:
            data = msg["data"]
            user.x = data["pos"]["x"]
            user.y = data["pos"]["y"]
            user.z = data["pos"]["z"]
            user.roll = data["ori"]["x"]
            user.yaw = data["ori"]["y"]
            user.pitch = data["ori"]["z"]
            user.state = data["state"]
            user.gesture = data["gesture"]
        await self._callback("on_user_pos", msg["user"], msg["data"])

    async def send(self, msg: dict) -> None:
        if self.ws is not None:
            await self.ws.send_message(json.dumps(msg))
            self.log(f'< {msg}')
        else:
            self.log('* Websocket not initialized')

    async def send_msg(self, msg: str) -> None:
        await self.send({'type': 'msg', 'data': msg})
        await self._callback('on_self_msg', msg)

    async def send_position(self) -> None:
        await self.send({'type': 'pos', 'data': {'pos': {'x': round(self.x, 2),
                                                         'y': round(self.y, 2),
                                                         'z': round(self.z, 2)},
                                                 'ori': {'x': round(self.roll),
                                                         'y': round(self.yaw),
                                                         'z': round(self.pitch)},
                                                 'state': self.state,
                                                 'gesture': self.gesture}})

    async def change_avatar(self, avatar: int) -> None:
        self.avatar = avatar
        await self.send({'type': 'avatar', 'data': self.avatar})

    async def reader(self, websocket) -> None:
        while True:
            try:
                message_raw = await websocket.get_message()
                msg = json.loads(message_raw)
                await self._process_msg(msg)
            except trio_websocket.ConnectionClosed:
                break

    async def login(self) -> None:
        rlogin = await asks.post(f'{self.web_url}/auth',
                                 json={'login': self.name, 'password': 'password'})
        self.cookiejar = {
            AUTH_COOKIE: get_cookie_from_response(rlogin, AUTH_COOKIE)
        }

    async def get_world_list(self) -> None:
        r_world_list = await asks.get(f'{self.web_url}/world/', cookies=self.cookiejar)
        self.worldlist.clear()
        for w in r_world_list.json():
            self.worldlist[w['id']] = {'name': w['name'], 'users': w['users']}

    async def world_enter(self, world_id) -> None:
        r_world = await asks.get(f'{self.web_url}/world/{world_id}', cookies=self.cookiejar)
        self.world = r_world.json()['id']

    async def connect(self) -> None:
        await self.login()
        await self.get_world_list()
        headers = [('Cookie', f"{AUTH_COOKIE}={self.cookiejar[AUTH_COOKIE]}")]
        async with trio_websocket.open_websocket_url(self.ws_url, extra_headers=headers) as websocket:
            self.ws = websocket
            self.log('@ Connected')
            self.connected = True
            await self._callback('on_connected')
            async with trio.open_nursery() as nursery:
                self.nursery = nursery
                nursery.start_soon(self.reader, websocket)
            self.connected = False
            self.log('@ Disconnected')
            await self._callback('on_disconnected')

    def run(self) -> None:
        trio.run(self.connect)

File: bot/src/bot.ts
import WebSocket from 'ws'
import axios, {AxiosResponse} from 'axios'

const AUTH_COOKIE = 'lemuria_token_access'
const DEBUG = false

interface UserData {
  id: string
  name: string
  avatar: number
  world: number
}

interface WorldData {
  id: number
  name: string
  users: number
}

class User {
  name: string
  world: number
  x: number
  y: number
  z: number
  roll: number
  yaw: number
  pitch: number
  avatar: number
  state: string
  gesture: string | null

  constructor(name: string = '') {
    this.name = name
    this.world = 0
    this.x = 0.0
    this.y = 0.0
    this.z = 0.0
    this.roll = 0.0
    this.yaw = 0.0
    this.pitch = 0.0
    this.avatar = 0
    this.state = 'idle'
    this.gesture = null
  }

  setPosition(
    x: number = 0.0,
    y: number = 0.0,
    z: number = 0.0,
    roll: number = 0.0,
    yaw: number = 0.0,
    pitch: number = 0.0
  ): void {
    this.x = x
    this.y = y
    this.z = z
    this.roll = roll
    this.yaw = yaw
    this.pitch = pitch
  }
}

class Bot extends User {
  webUrl: string
  wsUrl: string
  ws: WebSocket | null
  loggingEnabled: boolean
  connected: boolean
  handlers: Record<string, any>
  userlist: Record<string, User>
  worldlist: Record<number, {name: string; users: number}>
  cookiejar: Record<string, string>

  constructor(webUrl: string, wsUrl: string, loggingEnabled: boolean = true) {
    super()
    this.webUrl = webUrl
    this.wsUrl = wsUrl
    this.ws = null
    this.loggingEnabled = loggingEnabled
    this.connected = false
    this.handlers = {}
    this.userlist = {}
    this.worldlist = {}
    this.cookiejar = {}
  }

  log(txt: string): void {
    if (!this.loggingEnabled) {
      return
    }
    console.log(txt)
  }

  async _callback(name: string, ...parameters: any[]): Promise<void> {
    const instances = [this, ...Object.values(this.handlers)]
    for (const inst of instances) {
      const f = (inst as any)[name]
      if (typeof f !== 'function') continue
      if (DEBUG) {
        this.log(`calling ${name}() on instance ${inst}`)
      }
      if (f) {
        await f.apply(inst, parameters)
      }
    }
  }

  async _processMsg(msg: any): Promise<void> {
    this.log(`> ${JSON.stringify(msg)}`)

    if (!msg.type) {
      this.log('* unknown message')
      return
    }

    const t = msg.type

    switch (t) {
      case 'avatar':
        const user = this.userlist[msg.user]
        if (user) {
          user.avatar = msg.data
        }
        await this._callback('on_user_avatar', msg.user, msg.data)
        break
      case 'join':
        await this._callback('on_user_join', msg.data)
        break
      case 'list':
        this.userlist = {}
        for (const u of msg.data as UserData[]) {
          const user = new User(u.name)
          user.avatar = u.avatar
          user.world = u.world
          this.userlist[u.id] = user
        }
        await this._callback('on_user_list')
        break
      case 'msg':
        await this._callback('on_msg', msg.user, msg.data)
        break
      case 'part':
        await this._callback('on_user_part', msg.data)
        break
      case 'pos':
        const posUser = this.userlist[msg.user]
        if (posUser) {
          const data = msg.data
          posUser.x = data.pos.x
          posUser.y = data.pos.y
          posUser.z = data.pos.z
          posUser.roll = data.ori.x
          posUser.yaw = data.ori.y
          posUser.pitch = data.ori.z
          posUser.state = data.state
          posUser.gesture = data.gesture
        }
        await this._callback('on_user_pos', msg.user, msg.data)
        break
    }
  }

  async send(msg: any): Promise<void> {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(msg))
      this.log(`< ${JSON.stringify(msg)}`)
    } else {
      this.log('* WebSocket not initialized or not open')
    }
  }

  async sendMsg(msg: string): Promise<void> {
    await this.send({type: 'msg', data: msg})
    await this._callback('on_self_msg', msg)
  }

  async sendPosition(): Promise<void> {
    await this.send({
      type: 'pos',
      data: {
        pos: {
          x: Number(this.x.toFixed(2)),
          y: Number(this.y.toFixed(2)),
          z: Number(this.z.toFixed(2))
        },
        ori: {
          x: Math.round(this.roll),
          y: Math.round(this.yaw),
          z: Math.round(this.pitch)
        },
        state: this.state,
        gesture: this.gesture
      }
    })
  }

  async changeAvatar(avatar: number): Promise<void> {
    this.avatar = avatar
    await this.send({type: 'avatar', data: this.avatar})
  }

  async login(): Promise<void> {
    try {
      const response: AxiosResponse = await axios.post(`${this.webUrl}/auth/`, {
        login: this.name,
        password: 'password'
      })
      const cookie = response.headers['set-cookie']?.find((cookie) =>
        cookie.startsWith(AUTH_COOKIE)
      )
      if (cookie) {
        this.cookiejar[AUTH_COOKIE] = cookie.split(';')[0].split('=')[1]
      }
    } catch (error) {
      console.error('Login failed:', error)
    }
  }

  async getWorldList(): Promise<void> {
    try {
      const response: AxiosResponse<WorldData[]> = await axios.get(
        `${this.webUrl}/world/`,
        {
          headers: {Cookie: `${AUTH_COOKIE}=${this.cookiejar[AUTH_COOKIE]}`}
        }
      )
      this.worldlist = {}
      for (const w of response.data) {
        this.worldlist[w.id] = {name: w.name, users: w.users}
      }
    } catch (error) {
      console.error('Failed to get world list:', error)
    }
  }

  async worldEnter(worldId: number): Promise<void> {
    try {
      const response: AxiosResponse<{id: number}> = await axios.get(
        `${this.webUrl}/world/${worldId}`,
        {
          headers: {Cookie: `${AUTH_COOKIE}=${this.cookiejar[AUTH_COOKIE]}`}
        }
      )
      this.world = response.data.id
    } catch (error) {
      console.error('Failed to enter world:', error)
    }
  }

  async connect(): Promise<void> {
    await this.login()
    await this.getWorldList()

    this.ws = new WebSocket(this.wsUrl, {
      headers: {Cookie: `${AUTH_COOKIE}=${this.cookiejar[AUTH_COOKIE]}`}
    })

    this.ws.on('open', async () => {
      this.log('@ Connected')
      this.connected = true
      await this._callback('on_connected')
    })

    this.ws.on('message', async (data: WebSocket.Data) => {
      const msg = JSON.parse(data.toString())
      await this._processMsg(msg)
    })

    this.ws.on('close', async () => {
      this.connected = false
      this.log('@ Disconnected')
      await this._callback('on_disconnected')
    })

    this.ws.on('error', (error: Error) => {
      console.error('WebSocket error:', error)
    })
  }

  run(): void {
    this.connect().catch(console.error)
  }
}

export {User, Bot}
